sshd-core/src/main/java/org/apache/sshd/client/session/AbstractClientSession.java

```diff
@@ -231,37 +240,55 @@ public KeyPair removePublicKeyIdentity(KeyPair kp) {
             return null;
         }
 
-        int index = AuthenticationIdentitiesProvider.findIdentityIndex(
-            identities, AuthenticationIdentitiesProvider.KEYPAIR_IDENTITY_COMPARATOR, kp);
+        int index = AuthenticationIdentitiesProvider.findIdentityIndex(identities,
+                AuthenticationIdentitiesProvider.KEYPAIR_IDENTITY_COMPARATOR, kp);
         if (index >= 0) {
             return (KeyPair) identities.remove(index);
         } else {
             return null;
         }
     }
 
-    protected IoWriteFuture sendClientIdentification() throws Exception {
-        clientVersion = resolveIdentificationString(ClientFactoryManager.CLIENT_IDENTIFICATION);
+    protected void initializeKexPhase() throws Exception {
+        sendClientIdentification();
+
+        KexExtensionHandler extHandler = getKexExtensionHandler();
+        if ((extHandler == null) || (!extHandler.isKexExtensionsAvailable(this, AvailabilityPhase.PREKEX))) {
+            kexState.set(KexState.INIT);
+            sendKexInit();
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("initializeKexPhase({}) delay KEX-INIT until server-side one received", this);
+            }
+        }
+    }
 
+    protected ClientProxyConnector initializeProxyConnector() throws Exception {
```

Comment:
Why return the ClientProxyConnector here? The only place it's called the return value is ignored, and subclasses could get it via `getClientProxyConnector()` anyway.